base/linalg/adjtrans.jl

```diff
@@ -99,22 +119,100 @@ parent(A::AdjOrTrans) = A.parent
 vec(v::AdjOrTransAbsVec) = v.parent
 
 
+### concatenation
+# preserve Adjoint/Transpose wrapper around vectors
+# to retain the associated semantics post-concatenation
+hcat(avs::Union{Number,AdjointAbsVec}...) = _adjoint_hcat(avs...)
+hcat(tvs::Union{Number,TransposeAbsVec}...) = _transpose_hcat(tvs...)
+_adjoint_hcat(avs::Union{Number,AdjointAbsVec}...) = Adjoint(vcat(map(Adjoint, avs)...))
+_transpose_hcat(tvs::Union{Number,TransposeAbsVec}...) = Transpose(vcat(map(Transpose, tvs)...))
+typed_hcat(::Type{T}, avs::Union{Number,AdjointAbsVec}...) where {T} = Adjoint(typed_vcat(T, map(Adjoint, avs)...))
+typed_hcat(::Type{T}, tvs::Union{Number,TransposeAbsVec}...) where {T} = Transpose(typed_vcat(T, map(Transpose, tvs)...))
+# otherwise-redundant definitions necessary to prevent hitting the concat methods in sparse/sparsevector.jl
+hcat(avs::Adjoint{<:Any,<:Vector}...) = _adjoint_hcat(avs...)
+hcat(tvs::Transpose{<:Any,<:Vector}...) = _transpose_hcat(tvs...)
+hcat(avs::Adjoint{T,Vector{T}}...) where {T} = _adjoint_hcat(avs...)
+hcat(tvs::Transpose{T,Vector{T}}...) where {T} = _transpose_hcat(tvs...)
+
+
+### higher order functions
+# preserve Adjoint/Transpose wrapper around vectors
+# to retain the associated semantics post-map/broadcast
+
+# vectorfy takes an Adoint/Transpose-wrapped vector and builds
+# an unwrapped vector with the entrywise-same contents
+vectorfy(x::Number) = x
+vectorfy(adjvec::AdjointAbsVec) = map(Adjoint, adjvec.parent)
+vectorfy(transvec::TransposeAbsVec) = map(Transpose, transvec.parent)
+vectorfyall(transformedvecs...) = (map(vectorfy, transformedvecs)...,)
+
+# map over collections of Adjoint/Transpose-wrapped vectors
+# note that the caller's operation `f` should be applied to the entries of the wrapped
+# vectors, rather than the entires of the wrapped vector's parents. so first we use vectorfy
+# to build unwrapped vectors with entrywise-same contents as the wrapped input vectors.
+# then we map the caller's operation over that set of unwrapped vectors. but now re-wrapping
+# the resulting vector would inappropriately transform the result vector's entries. so
+# instead of simply mapping the caller's operation over the set of unwrapped vectors,
+# we map Adjoint/Transpose composed with the caller's operationt over the set of unwrapped
+# vectors. then re-wrapping the result vector yields a wrapped vector with the correct entries.
+map(f, avs::AdjointAbsVec...) = Adjoint(map(Adjoint∘f, vectorfyall(avs...)...))
```

Comment:
I don't understand why `getindex` doesn't just solve this for you. Are you trying to return an `Adjoint`-wrapped object? If so, why?

Reply:
Indeed, returning an `Adjoint`/`Transpose`-wrapped vector from higher order operations over `Adjoint`/`Transpose`-wrapped vectors is these methods' purpose :).

Why? The superficial reason is preservation of `ConjRowVector`/`RowVector`'s behavior in `Adjoint`/`Transpose`-wrapped vectors. The reason for `ConjRowVector`/`RowVector`'s behavior: Without re-wrapping, the result of these operations is a single-row `Matrix`, which has different semantics from `ConjRowVector`/`RowVector`. Consequently, for example, `foo.(anadjointvector) * avector` would produce a one-by-one matrix rather than a scalar.

In other words, re-wrapping preserves 4774 semantics through higher order operations insofar as possible.

Does that clarify? Thoughts? :)

Reply:
(Sidenote: typo in the comment `operationt`)

Reply:
Good catch! Thanks! :)

Reply:
What about `Adjoint(map(Adjoint∘f∘Adjoint, parent.(avs)...))`?

Reply:
(so you don't have to create a temporary)

Reply:
Avoiding the temporaries would be lovely! Deciphering the above expression is more than I have mental bandwidth for right at the moment though 😄. Perhaps a post-freeze optimization? :)

Reply:
@timholy, revisiting this rewrite briefly, given `Adjoint` is unary I imagine you instead meant e.g. `Adjoint(map((xs...) -> Adjoint(f(Adjoint.(xs)...)), parent.(avs)...))`? Thanks again! :)

Reply:
Yes, that looks right.